teps/0050-ignore-task-failures.md

```diff
@@ -0,0 +1,102 @@
+or two tasks are connected by an edge which is defined using either resource dependency (`from` or `task results`) or
+ordering dependency (`runAfter`). One single task failure results in a pipeline failure i.e. a task resulting in a
+failure blocks executing the rest of the graph. 
+
+```yaml
+$ kubectl get pr pipelinerun-with-failing-task-csmjr -o json | jq .status.conditions
+[
+  {
+    "lastTransitionTime": "2021-02-05T18:51:15Z",
+    "message": "Tasks Completed: 1 (Failed: 1, Cancelled 0), Skipped: 3",
+    "reason": "Failed",
+    "status": "False",
+    "type": "Succeeded"
+  }
+]
+```
+
+Tekton [catalog](https://github.com/tektoncd/catalog) has a wide range of `tasks` which are designed to be reusable
+in many pipelines. As a pipeline execution engine, we encourage the pipeline authors to utilize arbitrary tasks from
+the Tekton catalog. But, many common pipelines have the requirement where a task failure must not block executing the
+rest of the tasks.
+
+A pipeline author has an option to utilize `finally` section of the pipeline in which all the final tasks are executed
+after all the tasks in a graph have completed regardless of success or failure. `finally` has its own advantages and
+very helpful in various use cases including notifications, cleanup, etc.
+
+But, the pipeline authors does not have the flexibility to unblock executing the rest of the graph after experiencing a
+single task failure.
+
+
+## Motivation
+
+It should be possible to utilize tasks from the Tekton catalog in a pipeline. A pipeline author has no
+control over the task definitions but may desire to ignore a failure and continue executing the rest of the graph.
+
+
+### Goals
+
+* Design a task failure strategy so that the pipeline author can control the behavior of the underlying tasks 
+  and decide whether to continue executing the rest of the graph in the event of failure.
+
+* Be applicable to any pipeline with references to the tasks in a catalog or inlined task specifications.
+
+### Non-Goals
+
+* Not an alternative to combining the tasks in a pipeline which is covered in
+  [TEP-0044 Composing Tasks with Tasks](https://github.com/tektoncd/community/pull/316).
+
+* Not optimizing pipeline runtime which is covered in
+  [TEP-0046 PipelineRun in a Pod](https://github.com/tektoncd/community/pull/318).
+
+## Requirements
+
+* Users should be able to use any task from the catalog without having to alter its specification to allow that task to
+  fail without stopping the execution of a pipeline.
+
+* It should be possible to know that a task failed, and the rest of the graph was allowed to continue by observing
+  the status of the `PipelineRun`.
+
+
+### Use Cases
+
+* As a pipeline author, I would like to design a pipeline where a task running
+  [unit tests](https://github.com/tektoncd/catalog/tree/master/task/golang-test/0.1) might fail,
+  but can continue running integration tests, so that my pipeline can identify failures in both the tests.
```

Comment:
is it possible to explain a bit more here? one way this could be addressed today is to run the unit tests and integration tests in parallel; that way the unit tests can fail but the integration tests will still complete

Reply:
@pritidesai i was thinking about this more and i have a couple ideas of how we can tweak the use cases so that running the tests in parallel won't solve the problem (maybe this is what you already had in mind)

* A "warning only" use case - e.g. a pipeline wants to run some tests, and in parallel it wants to run something option, e.g. maybe some really intense linting, and the pipeline doesn't need to fail if the linting fails

```
    |             |
    v             v
linting       unit tests
                  |
                  v    
              integration tests
``` 

There's another very similar use case, where you have the same pipeline as above, but you DO want linting to fail the pipeline, you just don't want it to interrupt the unit tests -> integration tests branch <-- for that case i wonder if pipelines in pipelines is more appropriate? not sure!

Reply:
thanks @bobcatfish for looking into it 😻 

Yes, pipeline in pipeline can be an appropriate solution to not interrupt the combo of `unit tests -> integration tests`. I will add both the points (running in parallel and pipeline in pipeline) in the TEP.

But, I was thinking of a use case I don't want `linting` failure or `unit test` failure to interrupt deploying an application to a staging cluster so that the new features in the app can be shared for early feedback.

```
     |                |                      |
    v               v                     v
linting       unit tests      integration tests
                                            |
                                           v    
                                      deploying to staging cluster
```

One more use case I was thinking of having one single GitHub repo with multiple runtimes, and building/deploying an image for each runtime. and I don't want failure in one runtime interrupt the other runtime. But we can argue that this use case can fit well with `pipeline in pipeline` approach.

Reply:
> Do fail the pipeline if a task fail but do not interrupt other branches (or even the current one, also this would seem weird to me)

I agree - I think we're trying to tackle (1) and could leave (2) as separate (maybe pipelines in pipelines or something there is the best solution for (2))

Reply:
thanks @bobcatfish @vdemeester 
yup sure, I just updated the TEP to include only (1)